File: backend/app/services/rubros_gastos.py
from sqlalchemy import select, func, and_
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.rubros import RubroGasto
from app.models.cdp import CDP


async def get_rubros(db: AsyncSession, tenant_id: str, solo_hojas: bool = False) -> list[RubroGasto]:
    stmt = select(RubroGasto).where(RubroGasto.tenant_id == tenant_id)
    if solo_hojas:
        stmt = stmt.where(RubroGasto.es_hoja == 1)
    stmt = stmt.order_by(RubroGasto.codigo)
    result = await db.execute(stmt)
    return list(result.scalars().all())


async def get_rubro(db: AsyncSession, tenant_id: str, codigo: str) -> RubroGasto | None:
    result = await db.execute(
        select(RubroGasto).where(
            RubroGasto.tenant_id == tenant_id,
            RubroGasto.codigo == codigo,
        )
    )
    return result.scalar_one_or_none()


async def buscar(db: AsyncSession, tenant_id: str, filtro: str) -> list[RubroGasto]:
    stmt = select(RubroGasto).where(
        and_(
            RubroGasto.tenant_id == tenant_id,
            RubroGasto.es_hoja == 1,
            (RubroGasto.codigo.ilike(f"%{filtro}%") | RubroGasto.cuenta.ilike(f"%{filtro}%")),
        )
    ).order_by(RubroGasto.codigo)
    result = await db.execute(stmt)
    return list(result.scalars().all())


async def saldo_disponible(db: AsyncSession, tenant_id: str, codigo_rubro: str) -> float:
    rubro = await get_rubro(db, tenant_id, codigo_rubro)
    if not rubro:
        return 0
    stmt = select(func.coalesce(func.sum(CDP.valor), 0)).where(
        and_(
            CDP.tenant_id == tenant_id,
            CDP.codigo_rubro == codigo_rubro,
            CDP.estado != "ANULADO",
        )
    )
    result = await db.execute(stmt)
    total_cdps = result.scalar()
    return rubro.apropiacion_definitiva - total_cdps


# Alias usado en cdp.py
async def saldo_disponible_rubro(db: AsyncSession, tenant_id: str, codigo_rubro: str) -> float:
    return await saldo_disponible(db, tenant_id, codigo_rubro)


async def crear(db: AsyncSession, tenant_id: str, codigo: str, cuenta: str,
                apropiacion_definitiva: float = 0, apropiacion_inicial: float = 0) -> RubroGasto:
    existing = await get_rubro(db, tenant_id, codigo)
    if existing:
        raise ValueError(f"El rubro {codigo} ya existe")

    rubro = RubroGasto(
        tenant_id=tenant_id,
        codigo=codigo,
        cuenta=cuenta,
        es_hoja=1,
        apropiacion_inicial=apropiacion_inicial if apropiacion_inicial > 0 else apropiacion_definitiva,
        apropiacion_definitiva=apropiacion_definitiva,
    )
    db.add(rubro)

    # Mark parent as non-leaf
    parts = codigo.rsplit(".", 1)
    if len(parts) > 1:
        parent_code = parts[0]
        parent = await get_rubro(db, tenant_id, parent_code)
        if parent:
            parent.es_hoja = 0

    await db.commit()
    await _recalcular_hojas(db, tenant_id)
    return rubro


async def editar(db: AsyncSession, tenant_id: str, codigo: str, cuenta: str | None = None,
                 apropiacion_definitiva: float | None = None,
                 apropiacion_inicial: float | None = None) -> RubroGasto:
    rubro = await get_rubro(db, tenant_id, codigo)
    if not rubro:
        raise ValueError(f"Rubro {codigo} no encontrado")

    if cuenta is not None:
        rubro.cuenta = cuenta

    if apropiacion_definitiva is not None:
        rubro.apropiacion_definitiva = apropiacion_definitiva
        no_mods = (rubro.adiciones == 0 and rubro.reducciones == 0
                   and rubro.creditos == 0 and rubro.contracreditos == 0)
        if no_mods:
            rubro.apropiacion_inicial = apropiacion_definitiva

    if apropiacion_inicial is not None and apropiacion_definitiva is None:
        rubro.apropiacion_inicial = apropiacion_inicial
        rubro.apropiacion_definitiva = (
            apropiacion_inicial + rubro.adiciones - rubro.reducciones
            + rubro.creditos - rubro.contracreditos
        )

    await db.commit()
    await db.refresh(rubro)
    return rubro


async def eliminar(db: AsyncSession, tenant_id: str, codigo: str):
    rubro = await get_rubro(db, tenant_id, codigo)
    if not rubro:
        raise ValueError(f"Rubro {codigo} no encontrado")

    # Check for CDPs
    stmt = select(func.count()).select_from(CDP).where(
        CDP.tenant_id == tenant_id,
        CDP.codigo_rubro == codigo,
    )
    result = await db.execute(stmt)
    if result.scalar() > 0:
        raise ValueError("No se puede eliminar: tiene CDPs registrados")

    # Check for children
    stmt = select(func.count()).select_from(RubroGasto).where(
        RubroGasto.tenant_id == tenant_id,
        RubroGasto.codigo.like(f"{codigo}.%"),
    )
    result = await db.execute(stmt)
    if result.scalar() > 0:
        raise ValueError("No se puede eliminar: tiene sub-rubros")

    await db.delete(rubro)
    await db.commit()
    await _recalcular_hojas(db, tenant_id)


async def _recalcular_hojas(db: AsyncSession, tenant_id: str):
    all_rubros = await get_rubros(db, tenant_id)
    codigos = {r.codigo for r in all_rubros}
    for rubro in all_rubros:
        has_children = any(c.startswith(rubro.codigo + ".") for c in codigos if c != rubro.codigo)
        new_val = 0 if has_children else 1
        if rubro.es_hoja != new_val:
            rubro.es_hoja = new_val
    await db.commit()


async def sincronizar_padres(db: AsyncSession, tenant_id: str):
    all_rubros = await get_rubros(db, tenant_id)
    hojas = [r for r in all_rubros if r.es_hoja == 1]
    padres = [r for r in all_rubros if r.es_hoja == 0]

    for padre in padres:
        hijos = [h for h in hojas if h.codigo.startswith(padre.codigo + ".")]
        if hijos:
            padre.apropiacion_inicial = sum(h.apropiacion_inicial for h in hijos)
            padre.adiciones = sum(h.adiciones for h in hijos)
            padre.reducciones = sum(h.reducciones for h in hijos)
            padre.creditos = sum(h.creditos for h in hijos)
            padre.contracreditos = sum(h.contracreditos for h in hijos)
            padre.apropiacion_definitiva = sum(h.apropiacion_definitiva for h in hijos)
    await db.commit()

File: backend-cloudflare/src/routes/rubros-gastos.ts
/**
 * Rubros de Gastos - Endpoints
 * Migrado de Python/FastAPI a TypeScript/Hono
 */

import { Hono } from 'hono';
import { clerkAuth, requireEscritura } from '../middleware/auth';
import { getDb } from '../db';
import { rubrosGastos, cdp } from '../db/schema';
import type { Env, Variables } from '../types/bindings';
import { eq, and, like, or, sql } from 'drizzle-orm';
import { zValidator } from '@hono/zod-validator';
import { z } from 'zod';

const app = new Hono<{ Bindings: Env; Variables: Variables }>();

// ============================================================================
// SCHEMAS DE VALIDACIÓN
// ============================================================================

const createRubroGastoSchema = z.object({
  codigo: z.string().min(1).max(50),
  cuenta: z.string().min(1).max(500),
  apropiacionDefinitiva: z.number().default(0),
  apropiacionInicial: z.number().default(0),
});

const updateRubroGastoSchema = z.object({
  cuenta: z.string().min(1).max(500).optional(),
  apropiacionDefinitiva: z.number().optional(),
  apropiacionInicial: z.number().optional(),
});

// ============================================================================
// FUNCIONES AUXILIARES
// ============================================================================

/**
 * Calcula el saldo disponible de un rubro
 */
async function calcularSaldoDisponible(
  db: any,
  tenantId: string,
  codigoRubro: string
): Promise<number> {
  // Obtener el rubro
  const rubro = await db.query.rubrosGastos.findFirst({
    where: and(
      eq(rubrosGastos.tenantId, tenantId),
      eq(rubrosGastos.codigo, codigoRubro)
    ),
  });

  if (!rubro) return 0;

  // Sumar todos los CDPs activos (no anulados)
  const result = await db
    .select({
      total: sql<number>`COALESCE(SUM(${cdp.valor}), 0)`,
    })
    .from(cdp)
    .where(
      and(
        eq(cdp.tenantId, tenantId),
        eq(cdp.codigoRubro, codigoRubro),
        sql`${cdp.estado} != 'ANULADO'`
      )
    );

  const totalCdps = Number(result[0]?.total || 0);
  return rubro.apropiacionDefinitiva - totalCdps;
}

/**
 * Recalcula qué rubros son hojas y cuáles no
 */
async function recalcularHojas(db: any, tenantId: string): Promise<void> {
  // Obtener todos los rubros del tenant
  const todosLosRubros = await db.query.rubrosGastos.findMany({
    where: eq(rubrosGastos.tenantId, tenantId),
  });

  const codigos = new Set<string>(todosLosRubros.map((r: any) => r.codigo));

  // Para cada rubro, verificar si tiene hijos
  for (const rubro of todosLosRubros) {
    const tieneHijos = Array.from(codigos).some(
      (c) => c !== rubro.codigo && c.startsWith(rubro.codigo + '.')
    );

    const nuevoValor = tieneHijos ? 0 : 1;

    if (rubro.esHoja !== nuevoValor) {
      await db
        .update(rubrosGastos)
        .set({ esHoja: nuevoValor })
        .where(eq(rubrosGastos.codigo, rubro.codigo));
    }
  }
}

// ============================================================================
// ENDPOINTS
// ============================================================================

/**
 * GET /api/rubros-gastos
 * Listar todos los rubros de gastos (opcionalmente solo hojas)
 */
app.get('/', clerkAuth, async (c) => {
  const tenantId = c.get('tenantId');
  const db = getDb(c.env);
  const soloHojas = c.req.query('solo_hojas') === 'true';

  let conditions = [eq(rubrosGastos.tenantId, tenantId)];

  if (soloHojas) {
    conditions.push(eq(rubrosGastos.esHoja, 1));
  }

  const rubros = await db.query.rubrosGastos.findMany({
    where: and(...conditions),
    orderBy: (rubrosGastos, { asc }) => [asc(rubrosGastos.codigo)],
  });

  // Calcular saldo disponible para cada rubro
  const result = await Promise.all(
    rubros.map(async (rubro) => ({
      ...rubro,
      saldoDisponible: await calcularSaldoDisponible(db, tenantId, rubro.codigo),
    }))
  );

  return c.json(result);
});

/**
 * GET /api/rubros-gastos/buscar?q=...
 * Buscar rubros por código o cuenta
 */
app.get('/buscar', clerkAuth, async (c) => {
  const tenantId = c.get('tenantId');
  const q = c.req.query('q');

  if (!q || q.length < 1) {
    return c.json({ error: 'Parámetro de búsqueda requerido' }, 400);
  }

  const db = getDb(c.env);

  const rubros = await db.query.rubrosGastos.findMany({
    where: and(
      eq(rubrosGastos.tenantId, tenantId),
      eq(rubrosGastos.esHoja, 1),
      or(
        like(rubrosGastos.codigo, `%${q}%`),
        like(rubrosGastos.cuenta, `%${q}%`)
      )
    ),
    orderBy: (rubrosGastos, { asc }) => [asc(rubrosGastos.codigo)],
  });

  return c.json(rubros);
});

/**
 * GET /api/rubros-gastos/:codigo
 * Obtener un rubro específico por código
 */
app.get('/:codigo', clerkAuth, async (c) => {
  const tenantId = c.get('tenantId');
  const codigo = c.req.param('codigo');
  const db = getDb(c.env);

  const rubro = await db.query.rubrosGastos.findFirst({
    where: and(
      eq(rubrosGastos.tenantId, tenantId),
      eq(rubrosGastos.codigo, codigo)
    ),
  });

  if (!rubro) {
    return c.json({ error: 'Rubro no encontrado' }, 404);
  }

  // Calcular saldo disponible
  const saldoDisponible = await calcularSaldoDisponible(db, tenantId, codigo);

  return c.json({
    ...rubro,
    saldoDisponible,
  });
});

/**
 * POST /api/rubros-gastos
 * Crear un nuevo rubro de gasto
 */
app.post('/', clerkAuth, requireEscritura, zValidator('json', createRubroGastoSchema), async (c) => {
  const tenantId = c.get('tenantId');
  const data = c.req.valid('json');
  const db = getDb(c.env);

  // Verificar si el rubro ya existe
  const existing = await db.query.rubrosGastos.findFirst({
    where: and(
      eq(rubrosGastos.tenantId, tenantId),
      eq(rubrosGastos.codigo, data.codigo)
    ),
  });

  if (existing) {
    return c.json({ error: `El rubro ${data.codigo} ya existe` }, 400);
  }

  // Determinar apropiación inicial
  const apropiacionInicial =
    data.apropiacionInicial > 0 ? data.apropiacionInicial : data.apropiacionDefinitiva;

  // Crear el rubro
  const [nuevoRubro] = await db
    .insert(rubrosGastos)
    .values({
      tenantId,
      codigo: data.codigo,
      cuenta: data.cuenta,
      esHoja: 1,
      apropiacionInicial,
      apropiacionDefinitiva: data.apropiacionDefinitiva,
    })
    .returning();

  // Marcar el padre como no-hoja si existe
  const parts = data.codigo.split('.');
  if (parts.length > 1) {
    const parentCode = parts.slice(0, -1).join('.');
    const parent = await db.query.rubrosGastos.findFirst({
      where: and(
        eq(rubrosGastos.tenantId, tenantId),
        eq(rubrosGastos.codigo, parentCode)
      ),
    });

    if (parent) {
      await db
        .update(rubrosGastos)
        .set({ esHoja: 0 })
        .where(eq(rubrosGastos.codigo, parentCode));
    }
  }

  // Recalcular hojas
  await recalcularHojas(db, tenantId);

  return c.json(nuevoRubro, 201);
});

/**
 * PUT /api/rubros-gastos/:codigo
 * Editar un rubro existente
 */
app.put('/:codigo', clerkAuth, requireEscritura, zValidator('json', updateRubroGastoSchema), async (c) => {
  const tenantId = c.get('tenantId');
  const codigo = c.req.param('codigo');
  const data = c.req.valid('json');
  const db = getDb(c.env);

  // Verificar que el rubro existe
  const rubro = await db.query.rubrosGastos.findFirst({
    where: and(
      eq(rubrosGastos.tenantId, tenantId),
      eq(rubrosGastos.codigo, codigo)
    ),
  });

  if (!rubro) {
    return c.json({ error: `Rubro ${codigo} no encontrado` }, 404);
  }

  // Preparar objeto de actualización
  const updates: any = {};

  if (data.cuenta !== undefined) {
    updates.cuenta = data.cuenta;
  }

  if (data.apropiacionDefinitiva !== undefined) {
    updates.apropiacionDefinitiva = data.apropiacionDefinitiva;

    // Si no hay modificaciones, actualizar también la inicial
    const noMods =
      rubro.adiciones === 0 &&
      rubro.reducciones === 0 &&
      rubro.creditos === 0 &&
      rubro.contracreditos === 0;

    if (noMods) {
      updates.apropiacionInicial = data.apropiacionDefinitiva;
    }
  }

  if (data.apropiacionInicial !== undefined && data.apropiacionDefinitiva === undefined) {
    updates.apropiacionInicial = data.apropiacionInicial;
    updates.apropiacionDefinitiva =
      data.apropiacionInicial +
      rubro.adiciones -
      rubro.reducciones +
      rubro.creditos -
      rubro.contracreditos;
  }

  // Actualizar el rubro
  const [actualizado] = await db
    .update(rubrosGastos)
    .set(updates)
    .where(
      and(eq(rubrosGastos.tenantId, tenantId), eq(rubrosGastos.codigo, codigo))
    )
    .returning();

  return c.json(actualizado);
});

/**
 * DELETE /api/rubros-gastos/:codigo
 * Eliminar un rubro
 */
app.delete('/:codigo', clerkAuth, requireEscritura, async (c) => {
  const tenantId = c.get('tenantId');
  const codigo = c.req.param('codigo');
  const db = getDb(c.env);

  // Verificar que el rubro existe
  const rubro = await db.query.rubrosGastos.findFirst({
    where: and(
      eq(rubrosGastos.tenantId, tenantId),
      eq(rubrosGastos.codigo, codigo)
    ),
  });

  if (!rubro) {
    return c.json({ error: `Rubro ${codigo} no encontrado` }, 404);
  }

  // Verificar que no tenga CDPs registrados
  const cdpsCount = await db
    .select({ count: sql<number>`count(*)` })
    .from(cdp)
    .where(
      and(eq(cdp.tenantId, tenantId), eq(cdp.codigoRubro, codigo))
    );

  if (Number(cdpsCount[0]?.count || 0) > 0) {
    return c.json({ error: 'No se puede eliminar: tiene CDPs registrados' }, 400);
  }

  // Verificar que no tenga sub-rubros
  const subRubrosCount = await db
    .select({ count: sql<number>`count(*)` })
    .from(rubrosGastos)
    .where(
      and(
        eq(rubrosGastos.tenantId, tenantId),
        like(rubrosGastos.codigo, `${codigo}.%`)
      )
    );

  if (Number(subRubrosCount[0]?.count || 0) > 0) {
    return c.json({ error: 'No se puede eliminar: tiene sub-rubros' }, 400);
  }

  // Eliminar el rubro
  await db
    .delete(rubrosGastos)
    .where(
      and(eq(rubrosGastos.tenantId, tenantId), eq(rubrosGastos.codigo, codigo))
    );

  // Recalcular hojas
  await recalcularHojas(db, tenantId);

  return c.body(null, 204);
});

/**
 * POST /api/rubros-gastos/sincronizar-padres
 * Sincronizar valores de rubros padre sumando los de sus hijos
 */
app.post('/sincronizar-padres', clerkAuth, requireEscritura, async (c) => {
  const tenantId = c.get('tenantId');
  const db = getDb(c.env);

  // Obtener todos los rubros
  const todosLosRubros = await db.query.rubrosGastos.findMany({
    where: eq(rubrosGastos.tenantId, tenantId),
  });

  const hojas = todosLosRubros.filter((r) => r.esHoja === 1);
  const padres = todosLosRubros.filter((r) => r.esHoja === 0);

  // Para cada padre, sumar los valores de sus hijos
  for (const padre of padres) {
    const hijos = hojas.filter((h) => h.codigo.startsWith(padre.codigo + '.'));

    if (hijos.length > 0) {
      const suma = {
        apropiacionInicial: hijos.reduce((sum, h) => sum + h.apropiacionInicial, 0),
        adiciones: hijos.reduce((sum, h) => sum + h.adiciones, 0),
        reducciones: hijos.reduce((sum, h) => sum + h.reducciones, 0),
        creditos: hijos.reduce((sum, h) => sum + h.creditos, 0),
        contracreditos: hijos.reduce((sum, h) => sum + h.contracreditos, 0),
        apropiacionDefinitiva: hijos.reduce((sum, h) => sum + h.apropiacionDefinitiva, 0),
      };

      await db
        .update(rubrosGastos)
        .set(suma)
        .where(eq(rubrosGastos.codigo, padre.codigo));
    }
  }

  return c.json({ message: 'Padres sincronizados' });
});

export default app;
